client/web/src/enterprise/codeintel/configuration/CodeIntelConfigurationPage.tsx

```diff
@@ -0,0 +1,610 @@
+    const [editor, setEditor] = useState<editor.ICodeEditor>()
+    const infer = useCallback(() => editor?.setValue(inferredConfiguration || ''), [editor, inferredConfiguration])
+
+    const customToolbar = useMemo<{
+        saveToolbar: React.FunctionComponent<SaveToolbarProps & AutoIndexProps>
+        propsGenerator: SaveToolbarPropsGenerator<AutoIndexProps>
+    }>(
+        () => ({
+            saveToolbar: CodeIntelAutoIndexSaveToolbar,
+            propsGenerator: props => {
+                const mergedProps = {
+                    ...props,
+                    onInfer: infer,
+                    loading: inferredConfiguration === undefined,
+                    inferEnabled: !!inferredConfiguration && configuration !== inferredConfiguration,
+                }
+                mergedProps.willShowError = () => !mergedProps.saving
+                mergedProps.saveDiscardDisabled = () => mergedProps.saving || !dirty
+
+                return mergedProps
+            },
+        }),
+        [dirty, configuration, inferredConfiguration, infer]
+    )
+
+    return fetchError ? (
+        <ErrorAlert prefix="Error fetching index configuration" error={fetchError} />
+    ) : (
+        <>
+            {saveError && <ErrorAlert prefix="Error saving index configuration" error={saveError} />}
+
+            {configuration === undefined ? (
+                <LoadingSpinner className="icon-inline" />
+            ) : (
+                <DynamicallyImportedMonacoSettingsEditor
+                    value={configuration}
+                    jsonSchema={allConfigSchema}
+                    canEdit={true}
+                    onSave={save}
+                    saving={state === EditorState.Saving}
+                    height={600}
+                    isLightTheme={isLightTheme}
+                    history={history}
+                    telemetryService={telemetryService}
+                    customSaveToolbar={customToolbar}
+                    onDirtyChange={setDirty}
+                    onEditor={setEditor}
+                />
+            )}
+        </>
+    )
+}
+
+interface AutoIndexProps {
+    loading: boolean
+    inferEnabled: boolean
+    onInfer?: () => void
+}
+
+const CodeIntelAutoIndexSaveToolbar: React.FunctionComponent<SaveToolbarProps & AutoIndexProps> = ({
+    dirty,
+    loading,
+    saving,
+    error,
+    onSave,
+    onDiscard,
+    inferEnabled,
+    onInfer,
+    saveDiscardDisabled,
+}) => (
+    <SaveToolbar
+        dirty={dirty}
+        saving={saving}
+        onSave={onSave}
+        error={error}
+        saveDiscardDisabled={saveDiscardDisabled}
+        onDiscard={onDiscard}
+    >
+        {loading ? (
+            <LoadingSpinner className="icon-inline mt-2 ml-2" />
+        ) : (
+            inferEnabled && (
+                <button
+                    type="button"
+                    title="Infer index configuration from HEAD"
+                    className="btn btn-link"
```

Comment:
For consistency, let's use `Button` from the `wildcard` package everywhere.

```suggestion
                <Button
                    type="button"
                    title="Infer index configuration from HEAD"
                    variant="link"
```